rust/arrow/src/compute/boolean_kernels.rs

```diff
@@ -0,0 +1,145 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use packed_simd::u8x64;
+use std::slice::{from_raw_parts, from_raw_parts_mut};
+
+use crate::buffer::{Buffer, MutableBuffer};
+use crate::builder::{BufferBuilderTrait, UInt8BufferBuilder};
+
+/// SIMD accelerated version of bitwise binary operation for two `Buffer`'s.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn bitwise_bin_op_simd<F>(left: &Buffer, right: &Buffer, op: F) -> Buffer
+where
+    F: Fn(u8x64, u8x64) -> u8x64,
+{
+    let mut result = MutableBuffer::new(left.len()).with_bitset(left.len(), false);
+    let lanes = u8x64::lanes();
+    for i in (0..left.len()).step_by(lanes) {
+        let left_data =
```

Comment:
I just want to understand the approach here. Do we do the following:

SIMD_AND (left data, right data)  -> result as bitmask
SIMD_AND(left validity, right validity) -> result as bitmask

The first SIMD would have performed the operation on garbage data as well (where validity bit is 0) but that is fine as we avoid a branch and finally use validity to get the end result

SIMD_AND (result1, result 2)

I am not sure how exactly do we use validity in the below code

Reply:
Yes, that's correct.  

Our `BooleanArray` (bit packed) and `Bitmask` are both backed by a `Buffer`.  For other kernels like ADD, etc. we will reuse the SIMD_AND on the `Bitmask`.

Reply:
Thanks.